web/src/ee/features/billing/server/cloudBillingRouter.ts

```diff
@@ -233,102 +377,682 @@ export const cloudBillingRouter = createTRPCRouter({
+          stripeClient,
+          subscription,
+        );
+
+        await client.subscriptions.update(stripeSubscriptionId, {
+          items: [
+            {
+              id: planItem.id,
+              price: newProduct.default_price as string,
+              quantity: 1,
+            },
+          ],
+          proration_behavior: "always_invoice", // Immediately invoice proration
+          // Note: Usage-based charges continue to the end of the cycle by design
+          ...cancellationPayload,
         });
 
-      await stripeClient.subscriptions.update(stripeSubscriptionId, {
-        items: [
-          // remove current product from subscription
+        return;
+      }
+
+      // Downgrade: schedule the price change at end of current billing period and
+      // keep the user on the old plan until then. Leave usage item running.
+
+      const currentPeriodEndSec =
+        (subscription as any).current_period_end ?? planItem.current_period_end;
+
+      // Build items arrays for the schedule phases
+
+      const nextPhaseItems = subscription.items.data.map((i) => {
+        const isMetered = i.price.recurring?.usage_type === "metered";
+        if (i.id === planItem.id) {
+          return { price: newProduct.default_price as string, quantity: 1 };
+        }
+        return {
+          price: i.price.id,
+          ...(isMetered ? {} : { quantity: i.quantity ?? 1 }),
+        };
+      });
+
+      // 1. Clear any existing schedules before creating a new one
+      await releaseExistingSubscriptionScheduleIfAny(
+        stripeClient,
+        subscription,
+      );
+
+      // 2. First create the schedule off the current subscription
+      const initialSchedule = await client.subscriptionSchedules.create({
+        from_subscription: stripeSubscriptionId,
+      });
+
+      // 3. Then update the schedule to the new plan
+      await client.subscriptionSchedules.update(initialSchedule.id, {
+        end_behavior: "release",
+        phases: [
+          // Keep the existing phase but set the end date to current period end
           {
-            id: item.id,
-            deleted: true,
+            start_date: initialSchedule.phases[0].start_date,
+            end_date: currentPeriodEndSec,
+            items: initialSchedule.phases[0]!.items as any, // ts error
+            proration_behavior: "none",
           },
-          // add new product to subscription
+          // Add new phase with the product change
           {
-            price: newProduct.default_price as string,
+            start_date: currentPeriodEndSec,
+            end_date: currentPeriodEndSec + 60, // End 60 seconds after the switch to release the schedule
+            items: nextPhaseItems,
+            proration_behavior: "none",
           },
         ],
-        // reset billing cycle which causes immediate invoice for existing plan
-        billing_cycle_anchor: "now",
-        proration_behavior: "none",
+        metadata: {
```

Comment:
pass through the zod object or at least set as const and use typescript typecheck against zod inferred ts type?